nixos/modules/services/security/bitwarden-directory-connector.nix

```diff
@@ -0,0 +1,346 @@
+        type = types.bool;
+        default = false;
+        description = lib.mdDoc "Enable if you are syncing more than 2000 users/groups.";
+      };
+
+      memberAttribute = mkOption {
+        type = types.str;
+        description = lib.mdDoc "Attribute that lists members in a LDAP group.";
+        example = "uniqueMember";
+      };
+
+      creationDateAttribute = mkOption {
+        type = types.str;
+        description = lib.mdDoc "Attribute that lists a users creation date.";
+        example = "whenCreated";
+      };
+
+      emailPrefixSuffix = {
+        enable = mkOption {
+          type = types.bool;
+          default = false;
+          description = lib.mdDoc "If a user has no email address, combine a username prefix with a suffix value to form an email.";
+        };
+        prefixAttribute = mkOption {
+          type = types.str;
+          description = lib.mdDoc "Attribute that has a users username.";
+          example = "accountName";
+        };
+        suffix = mkOption {
+          type = types.str;
+          description = lib.mdDoc "Suffix for the email, normally @example.com.";
+          example = "@example.com";
+        };
+      };
+
+      users = {
+        enable = mkOption {
+          type = types.bool;
+          default = false;
+          description = lib.mdDoc "Sync users.";
+        };
+        path = mkOption {
+          type = types.str;
+          description = lib.mdDoc "User directory, relative to root.";
+          example = "ou=users";
+        };
+        objectClass = mkOption {
+          type = types.str;
+          description = lib.mdDoc "A class that users will have.";
+          example = "inetOrgPerson";
+        };
+        emailAttribute = mkOption {
+          type = types.str;
+          description = lib.mdDoc "Attribute for a users email.";
+          example = "mail";
+        };
+        filter = mkOption {
+          type = types.str;
+          description = lib.mdDoc "Filter for users.";
+          example = "(memberOf=cn=sales,ou=groups,dc=example,dc=com)";
+        };
+      };
+      groups = {
+        enable = mkOption {
+          type = types.bool;
+          default = false;
+          description = lib.mdDoc "Sync groups.";
+        };
+        path = mkOption {
+          type = types.str;
+          description = lib.mdDoc "Group directory, relative to root.";
+          example = "ou=groups";
+        };
+        objectClass = mkOption {
+          type = types.str;
+          description = lib.mdDoc "A class that groups will have.";
+          example = "groupOfNames";
+        };
+        nameAttribute = mkOption {
+          type = types.str;
+          description = lib.mdDoc "Attribute for a name of group.";
+          example = "cn";
+        };
+        filter = mkOption {
+          type = types.str;
+          description = lib.mdDoc "Filter for groups.";
+          example = "(cn=sales)";
+        };
+      };
+    };
```

Comment:
This seems to scream for a `settings`-style option or three which just gets thrown at `builtins.toJSON`, unless you really want to type-verify every last bit of this for some reason. It would be more future-proof and flexible if users could add any json data they wanted to these data structures.

Reply:
Could ye expand a tad on this?  
I dont fully grasp what ye are saying.

Reply:
You've recreated an entire schema for a json datastructure here, which you aren't really looking inside, but just passing on near-verbatim to the underlying program. If the underlying program adds a field, your schema has to be updated to match that, or people can't use the module with the new feature. Why not just handle the whole datastructure as a single object?

What I'm suggesting is having 2 options, `sync` and `ldap`, both of type `pkgs.formats.json {}`, which are simply passed unmodified to `builtins.toJSON` to create `sync_data` and `ldap_data`.

If you want to document the available fields that people would often use, you can follow the example of `nix.settings` (defined [here](https://github.com/NixOS/nixpkgs/blob/nixos-23.05/nixos/modules/config/nix.nix#L169-L171)) and use a submodule with `freeformType`, so that you can list some suboptions, but the user can add ones you haven't listed.

You will need to clean up a couple of things to do this, like moving `sync.interval` out of the `sync` section, since you're treating it specially and not passing it through. The benefit, however, is that users can relate the options more clearly to upstream documentation, and can use fields that you either didn't know about, or that didn't exist when the module was created.

Reply:
I see.

I can see that working really well for ``sync``/``sync_data``, however ``ldap_data`` contains a password, would not be wise to write that to the nix store.

I'll take a gender at it.

Reply:
Proper handling of a password requires the ability to read it from a path at runtime. That would allow integration with secret systems like sops-nix and agenix, and keep it out of the store.

It doesn't necessarily invalidate using this scheme for `ldap_data`, though, it just means you need to shove an extra attribute in there with `jq` in `ExecStartPre` (which you needed to do anyway, to properly handle secrets).

Reply:
Just realised that I cna do ``ldap_data`` jsut fine as it gets set by the env var regardless.